### ISSUE TITLE
Prevent using `Error` when doing type-checking

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#27]
I often forget to use a `TypeError` when doing type-checking in a module.

Example:

``` js
if (Array.isArray(foo)) {
    throw new Error('Expected an array');
}
```

Would be nice to error the conditional only does type-checking and prevent the use of `Error`. Custom errors should be allowed.
## 

New to making ESLint rules? Check out this quick guide on [how to get started](https://gist.github.com/sindresorhus/1656c46f23545deff8cc713649dcff26).

**Comment by jamestalmage** [comment]
List of typechecking constructs:

- `Array.isArray`
- `Buffer.isBuffer`
- `typeof obj === desired`

**Comment by sindresorhus** [comment]
`instanceof` too, right?

**Comment by jamestalmage** [comment]
Yep. Probably others. Somebody should scan the JS API looking for them.

**Comment by jamestalmage** [comment]
- [`Function.prototype.isGenerator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/isGenerator)

- `isNan()`
- `isFinite()`
- Use of [`Symbol.species`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/species) - (I don't fully understand this yet, so not sure)
- Popular / obvious `isXXX` constructs (allowing for `lodash`, etc): 
  - `isString(val)` 
  - `isFunction(val)`
  - `isNumber` 
  - `isBoolean`
  - `isInteger`
  - `isPromise`
  - `isObservable`
  - `utils` / `lodash` type usage for all the above(`x.isString(val)` ...)

**Comment by florianb** [comment]
Okeydo - anybody working on this?

**Comment by sindresorhus** [comment]
@florianb Not that I know of. Go ahead :)

**Comment by florianb** [comment]
Hoorayy.. Hacktoberfest for the win! 🍺

**Comment by florianb** [comment]
How deep should i look for typechecks? How about multiple checks (`foo.isArray() || foo.isBuffer()`)?

**Comment by florianb** [comment]
@sindresorhus, @jamestalmage: I have a working draft now and i am ready to implement the different expressions.

The question is which standard(s) should i consider? Only official standards like the ES5/6/7 specifications?

I have also a question about connected style-issues like f.e. the precedence of expressions in BinaryExpressions. I decided to test without checking if the literal is written before the unary expression (like `'object' === typeof foo`), assuming this is handled by another rule.

But another case could be, that f.e. a type-checking expressions should probably not being combined with other expressions, if the only intention for the comparison is to throw an Error afterwards. F.e.

```
if (typeof foo !== 'object' || foo.bar() === false) {
  throw new Error('Expected Foo being bar!');
}
```

These statements should probable being split up then, don't they?

**Comment by sindresorhus** [comment]
> The question is which standard(s) should i consider? Only official standards like the ES5/6/7 specifications?

The latest. ES2017.

> I have also a question about connected style-issues like f.e. the precedence of expressions in BinaryExpressions. I decided to test without checking if the literal is written before the unary expression (like 'object' === typeof foo), assuming this is handled by another rule.

It should be handled. Rules should not be dependent on each other.

> But another case could be, that f.e. a type-checking expressions should probably not being combined with other expressions, if the only intention for the comparison is to throw an Error afterwards. F.e.

In that case it's ambiguous and we don't report anything.

**Comment by florianb** [comment]
Thanks!
